membership-attribute-service/test/services/PaymentFailureAlerterTest.scala

```diff
@@ -164,62 +169,172 @@ class PaymentFailureAlerterTest(implicit ee: ExecutionEnv)  extends Specificatio
-
-      lastUnpaidInvoiceDate === Some(invoiceDateOlder)
+    "mostRecentPayableInvoicesOlderThanOneMonth" should {
+
+      import InvoiceAndPaymentTestData._
+
+      "return an empty list if the user's invoices all have amount = 0" in {
+        val result = PaymentFailureAlerter.mostRecentPayableInvoicesOlderThanOneMonth(
+          List(
+            oldFreeInvoice,
+            oldFreeInvoice.copy(invoiceDate = moreThanTwoMonthsAgo, dueDate = moreThanTwoMonthsAgo),
+            oldFreeInvoice.copy(invoiceDate = moreThanThreeMonthsAgo, dueDate = moreThanThreeMonthsAgo)
+          )
+        )
+        result === Nil
+      }
+
+      "return an empty list if the user only has a new invoice" in {
+        val result = PaymentFailureAlerter.mostRecentPayableInvoicesOlderThanOneMonth(List(recentUnpaidInvoice))
+        result === Nil
+      }
+
+      "return the most recent two invoices older than one month if they are present" in {
+
+        val twoMonthOldInvoice = oldUnpaidInvoice.copy(invoiceDate = moreThanTwoMonthsAgo, dueDate = moreThanTwoMonthsAgo)
+        val threeMonthOldInvoice = oldUnpaidInvoice.copy(invoiceDate = moreThanThreeMonthsAgo, dueDate = moreThanThreeMonthsAgo)
+
+        val result = PaymentFailureAlerter.mostRecentPayableInvoicesOlderThanOneMonth(
+          List(
+            oldUnpaidInvoice,
+            twoMonthOldInvoice,
+            threeMonthOldInvoice
+          )
+        )
+        result === List(oldUnpaidInvoice, twoMonthOldInvoice)
+      }
+
+    }
+
+    "accountHasMissedRecentPayments" should {
+
+      import InvoiceAndPaymentTestData._
+
+      val accountId = AccountId("id123")
+
+      "return false if the user has a free sub" in {
+        PaymentFailureAlerter.accountHasMissedRecentPayments(accountId, isPaidSub = false, List(), List()) === false
+      }
+
+      "return false if the user has 0 amount invoices and no payments (i.e. sub which should be paid but has 100% discount)" in {
+        PaymentFailureAlerter.accountHasMissedRecentPayments(accountId, isPaidSub = true, List(oldFreeInvoice), List()) === false
+      }
+
+      "return false if the user has no payments or invoices for a paid sub (e.g. brand new sub)" in {
+        PaymentFailureAlerter.accountHasMissedRecentPayments(accountId, isPaidSub = true, List(), List()) === false
+      }
+
+      "return false if the user has a single paid invoice" in {
+        PaymentFailureAlerter.accountHasMissedRecentPayments(accountId, isPaidSub = true, List(paidInvoice), List(paymentForPaidInvoice)) === false
+      }
+
+      "return false if the user has a recent unpaid invoice" in {
+        PaymentFailureAlerter.accountHasMissedRecentPayments(accountId, isPaidSub = true, List(recentUnpaidInvoice), List()) === false
+        PaymentFailureAlerter.accountHasMissedRecentPayments(accountId, isPaidSub = true, List(recentUnpaidInvoice), List(failedPaymentForRecentUnpaidInvoice)) === false
```

Comment:
do you test 2 cases here on purpose ?

Reply:
Yes, I wanted to capture the case where we tried to take a payment and it failed vs the case where we didn't try to take a payment at all. Perhaps I should call it out as two separate scenarios?

Reply:
I think it would be slightly better to have one assertion per test case, but really I was just asking in case it was a copy and paste error or something. I'm fine with leaving it as it is

Reply:
> it would be slightly better to have one assertion per test case

Agreed, have fixed this.